<input type=checkbox switch> paints outside its box and the thumb is slightly misaligned on macOS

annevk opened the pull request:
#### 5381f48e7a048d58bf8b46f15a4318f38457acc8
<pre>
&lt;input type=checkbox switch&gt; paints outside its box and the thumb is slightly misaligned on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=267679">https://bugs.webkit.org/show_bug.cgi?id=267679</a>
<a href="https://rdar.apple.com/121579531">rdar://121579531</a>

Reviewed by Aditya Keerthi.

This fixes two problems:

1. The thumb was misplaced by half a pixel. The error here was not
   calculating the inflated box for the thumb independently and instead
   assuming it would be the inflated track&apos;s height squared. The new
   code calculates the thumb boxes (mostly) independently from the
   track boxes.

2. Drawing the thumb was not masked in any way. This resulted in some
   pixel bleeding at the top of the control. This was noticed, but
   initially not deemed a problem. It&apos;s now deemed a problem and
   corrected by using the same masking code we use for the track. There
   remains some slight bleeding, though noticeably less.

* LayoutTests/fast/forms/switch/no-pixels-outside-the-box-vertical-lr-rtl.html:
* LayoutTests/fast/forms/switch/no-pixels-outside-the-box.html:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
(WebCore::SwitchMacUtilities::visualCellSize):
(WebCore::SwitchMacUtilities::trackMaskImage):
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::cellSize const):
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::SwitchTrackMac::draw):
(WebCore::trackMaskImage): Deleted.

Canonical link: <a href="https://commits.webkit.org/275808@main">https://commits.webkit.org/275808@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34cb9e34c2b6c36e46e0e62d400a5bfb0d2362ee

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42876 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21897 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45277 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45494 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39001 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45182 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25596 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19277 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35482 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43449 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18925 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36911 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16454 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16535 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37970 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/932 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39056 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38304 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47016 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17703 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14596 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42229 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19328 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37239 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40887 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9567 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19492 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18958 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
```diff
@@ -29,7 +29,7 @@
 
 IntSize cellSize(NSControlSize);
-FloatSize visualCellSize(NSControlSize, const ControlStyle&);
+FloatSize visualCellSize(IntSize, const ControlStyle&);
```
Can you explain the motivation behind the type change? It seems like we simply moved the call to `cellSize` out of the method, but using the same `NSControlSize` and `ControlStyle`?

annevk replied:
It did not seem sound for visualCellSize() to call cellSize() directly as that would bypass the class method. And also, for the thumb we want to invoke cellSize() before visualCellSize() as it's easier to create the thumb variant from the logical size.

pxlcoder replied:
I see what you mean about the thumb 👍

annevk committed:
<input type=checkbox switch> paints outside its box and the thumb is slightly misaligned on macOS
https://bugs.webkit.org/show_bug.cgi?id=267679
rdar://121579531

Reviewed by Aditya Keerthi.

This fixes two problems:

1. The thumb was misplaced by half a pixel. The error here was not
   calculating the inflated box for the thumb independently and instead
   assuming it would be the inflated track's height squared. The new
   code calculates the thumb boxes (mostly) independently from the
   track boxes.

2. Drawing the thumb was not masked in any way. This resulted in some
   pixel bleeding at the top of the control. This was noticed, but
   initially not deemed a problem. It's now deemed a problem and
   corrected by using the same masking code we use for the track. There
   remains some slight bleeding, though noticeably less.

* LayoutTests/fast/forms/switch/no-pixels-outside-the-box-vertical-lr-rtl.html:
* LayoutTests/fast/forms/switch/no-pixels-outside-the-box.html:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
(WebCore::SwitchMacUtilities::visualCellSize):
(WebCore::SwitchMacUtilities::trackMaskImage):
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::cellSize const):
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::SwitchTrackMac::draw):
(WebCore::trackMaskImage): Deleted.

Canonical link: https://commits.webkit.org/275808@main
